src/target/xtensa/xtensa_chip.c

```diff
@@ -0,0 +1,252 @@
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.           *
+ ***************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "assert.h"
+#include <target/target.h>
+#include <target/target_type.h>
+#include <target/arm_adi_v5.h>
+#include <rtos/rtos.h>
+#include "xtensa_chip.h"
+#include "xtensa_fileio.h"
+
+
+int xtensa_chip_init_arch_info(struct target *target, void *arch_info, 
+		struct xtensa_debug_module_config *dm_cfg)
+{
+	struct xtensa_chip_common *xtensa_chip = (struct xtensa_chip_common *)arch_info;
+	if ((dm_cfg->queue_tdi_idle == NULL) && (dm_cfg->tap != NULL)) {
+		dm_cfg->queue_tdi_idle = xtensa_chip_queue_tdi_idle;
+		dm_cfg->queue_tdi_idle_arg = target;
+	}
+	int ret = xtensa_init_arch_info(target, &xtensa_chip->xtensa, dm_cfg);
+	if (ret != ERROR_OK) {
+		return ret;
+	}
+	/* All xtensa target structures point back to original xtensa_chip */
+	xtensa_chip->xtensa.xtensa_chip = arch_info;
+	return ERROR_OK;
+}
+
+static int xtensa_chip_handle_target_event(struct target *target, enum target_event event, void *priv)
+{
+	if (target != priv)
+		return ERROR_OK;
+
+	LOG_DEBUG("%d", event);
+
+	int ret = xtensa_handle_target_event(target, event, priv);
+	if (ret != ERROR_OK)
+		return ret;
+
+	switch (event) {
+	case TARGET_EVENT_HALTED:
+		/* debug stubs not supported */
+		break;
+	case TARGET_EVENT_GDB_DETACH:
+		/* Flash BP not supported */
+		break;
+	default:
+		break;
+	}
+	return ERROR_OK;
+}
+
+int xtensa_chip_target_init(struct command_context *cmd_ctx, struct target *target)
+{
+	int ret = xtensa_target_init(cmd_ctx, target);
+	if (ret != ERROR_OK)
+		return ret;
+
+	ret = target_register_event_callback(xtensa_chip_handle_target_event, target);
+	if (ret != ERROR_OK)
+		return ret;
+
+	ret = xtensa_fileio_init(target);
+	if (ret != ERROR_OK)
+		return ret;
+
+	return ERROR_OK;
+}
+
+int xtensa_chip_arch_state(struct target *target)
+{
+	return ERROR_OK;
+}
+
+static int xtensa_chip_poll(struct target *target)
+{
+	enum target_state old_state = target->state;
+	int ret;
+
+	ret = xtensa_poll(target);
+
+	if (old_state != TARGET_HALTED && target->state == TARGET_HALTED) {
+		/*Call any event callbacks that are applicable */
+		if (old_state == TARGET_DEBUG_RUNNING)
+			target_call_event_callbacks(target, TARGET_EVENT_DEBUG_HALTED);
+		else {
+			xtensa_fileio_detect_proc(target);
```

Comment:
Can we add fileio feature as a seperate patch? I have one semihosting patch also. Want to test this with that patch.

Reply:
Sure, I will separate it.